### PR TITLE
CLI: Adds info command

### PR DESCRIPTION
**Pull request by SM43**
image

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

### CODE REVIEW

File: api/pkg/cli/cmd/install/install_test.go [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+			},
+			&res.Tag{
+				ID:   1,
+				Name: "tag1",
+			},
+		},
+	},
+}
+
+// func TestCreateResource(t *testing.T) {
```
Comment (sthaha) [inline api/pkg/cli/cmd/install/install_test.go:91]: 🤔 did you meant to add a test?

File: api/pkg/cli/formatter/field.go [hunk 2]
```diff
@@ -52,3 +53,89 @@ func FormatTags(tags []*client.TagResponseBody) string {
...
+}
+
+func findSpaceIndexFromLast(str string) int {
+	return strings.LastIndex(str, " ")
+}
+
+// Icon returns icon for a title passed
+func Icon(icon string) string {
+
+	switch icon {
```
Comment (sthaha) [inline api/pkg/cli/formatter/field.go:106]: a `map[string][string]` will be simpler

File: api/pkg/cli/formatter/field.go [hunk 3]
```diff
@@ -52,3 +53,89 @@ func FormatTags(tags []*client.TagResponseBody) string {
 	}
 	return sb.String()
 }
+
+// FormatInfoDesc returns first 100 char of resource description
+func FormatInfoDesc(desc string) string {
```
Comment (sthaha) [inline api/pkg/cli/formatter/field.go:58]: Why is this `specific` to description? Aren't all formatters supposed to be generic?

Comment (sthaha) [inline api/pkg/cli/formatter/field.go:58]: How about calling it `WrapText(text string, maxWidth int) string` ?

File: api/pkg/cli/formatter/field.go [hunk 4]
```diff
@@ -52,3 +53,89 @@ func FormatTags(tags []*client.TagResponseBody) string {
 	}
 	return sb.String()
 }
+
+// FormatInfoDesc returns first 100 char of resource description
```
Comment (sthaha) [inline api/pkg/cli/formatter/field.go:57]: is this right anymore?

File: api/pkg/cli/formatter/field.go [hunk 5]
```diff
@@ -52,3 +53,89 @@ func FormatTags(tags []*client.TagResponseBody) string {
...
+func FormatInfoDesc(desc string) string {
+	defaultWidth := 80
```
Comment (sthaha) [inline api/pkg/cli/formatter/field.go:59]: IMHO, the caller should pass this value

File: api/pkg/cli/formatter/field.go (hunk 3 above)
Comment (sthaha) [reply api/pkg/cli/formatter/field.go:58]: Could you please add unit tests for all non-trivial formatters?

File: api/pkg/cli/formatter/field.go [hunk 6]
```diff
@@ -52,3 +53,89 @@ func FormatTags(tags []*client.TagResponseBody) string {
...
+// DefaultValue returns default value if string is empty
+func DefaultValue(val, def string) string {
+	if val == "" {
+		return def
+	}
+	return val
+}
+
+// ToLower returns string in lowercase
+func ToLower(str string) string {
```
Comment (sthaha) [inline api/pkg/cli/formatter/field.go:139]: why do we need this? can't we use `strings.ToLower` directly?

File: api/pkg/cli/hub/get_resource.go [hunk 7]
```diff
@@ -112,3 +140,46 @@ func (rr *ResourceResult) Manifest() ([]byte, error) {
 
 	return data, nil
 }
+
+// UnstructuredData returns resource as Unstructured object
+func (rr *ResourceResult) UnstructuredData() (*unstructured.Unstructured, error) {
```
Comment (sthaha) [inline api/pkg/cli/hub/get_resource.go:145]: do we have tests for this?

Comment (sthaha) [reply api/pkg/cli/hub/get_resource.go:145]: IMO `ToUnstructured()` or `Unstructured()` may read better

File: api/pkg/cli/cmd/info/info.go [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+	})
+
+	resource, err := res.Resource()
+	if err != nil {
+		return err
+	}
+
+	out := opts.cli.Stream().Out
+
+	if opts.version != "" {
```
Comment (piyush-garg) [inline api/pkg/cli/cmd/info/info.go:164]: Is not better to just return the resource based on the version asked, we can remove the if/else thing from the template then

Comment (SM43) [reply api/pkg/cli/cmd/info/info.go:143]: updated

File: api/pkg/cli/hub/get_resource.go [hunk 9]
```diff
@@ -33,28 +33,39 @@ type ResourceOption struct {
...
-	data, status, err := h.Get(opt.Endpoint())
+func (c *client) GetResource(opt ResourceOption) ResourceResult {
+	data, status, err := c.Get(opt.Endpoint())
 
 	return ResourceResult{
 		data:    data,
 		version: opt.Version,
 		status:  status,
 		err:     err,
+		set:     false,
```
Comment (piyush-garg) [inline api/pkg/cli/hub/get_resource.go:68]: What's the use of this?

**Comment by PuneetPunamiya** [comment]
Thanks for the patch, pr looks fine to me :call_me_hand: 
/lgtm
/approve
/meow

**Pull request merged**